crates/sui-graphql-rpc/schema.graphql

```diff
@@ -0,0 +1,840 @@
+
+  # (x)or any of these
+  sender: SuiAddress
+  transaction: Digest
+
+  # Cascading (module requires package)
+  emittingPackage: SuiAddress
+  emittingModule: String
+
+  # Cascading (type requires module requires package)
+  eventPackage: SuiAddress
+  eventModule: String
+  eventType: String
+
+  eventField: EventField
+
+  startTimeMs: U64
+  endTimeMs: U64
+}
+
+input EventField {
+  path: String!
+  value: JSON!
+}
+
+input TransactionBlockFilter {
+  # one of these
+  any: [TransactionBlockFilter]
+  all: [TransactionBlockFilter]
+  not: TransactionBlockFilter
+
+  # (x)or any of these
+
+  # Cascading (function requires module requires package)
+  package: SuiAddress
+  module: String
+  function: String
+
+  kind: TransactionBlockKindInput
+  checkpoint: U64
+  fromAddress: SuiAddress
+  toAddress: SuiAddress
+
+  inputObject: SuiAddress
+  changedObject: SuiAddress
+}
+
+type ConsistentRange {
+  first: Checkpoint
+  last: Checkpoint
+}
+
+interface Owner {
+  id: SuiAddress!
+
+  objects(query: ObjectQuery): ObjectQueryConnection
+
+  balance(type: TypeInst!): Balance
+  balances: BalanceConnection
+
+  coins(type: TypeInst): CoinConnection
+
+  stakes: StakeConnection
+
+  defaultNameServiceName: String
+  nameServiceNames: NameServiceNamesConnection
+}
+
+# Returned by Object.owner, where we can't disambiguate between
+# Address and Object.
+type AmbiguousOwner implements Owner {}
+
+type Address implements Owner {
+  transactionBlocks(
+    filter: AddressTransactionBlockFilter,
+  ): TransactionBlockConnection
+}
+
+enum AddressTransactionBlockFilter {
+  SIGN # Transactions this address has signed
+  SENT # Transactions that transferred objects from this address
+  RECV # Transactions that received objects into this address
+  PAID # Transactions that were paid for by this address
+}
+
+type Object implements Owner {
+  owner: AmbiguousOwner
+  version: U64
+  digest: Digest
+
+  previousVersion: U64
+  previousTransactionBlock: TransactionBlock
+  storageRebate: U64
+
+  display: [DisplayEntry]
+
+  asMoveObject: MoveObject
+  asMovePackage: MovePackage
+
+  # Transaction Blocks that sent objects to this object
+  receivedTransactionBlocks: TransactionBlockConnection
+
+  dynamicField(dynamicFieldName: DynamicFieldName!): DynamicField
+  dynamicFields: DynamicFieldConnection
+
+  bcs: Bytes
+}
+
+type DisplayEntry {
+  key: String
+  value: String
+}
+
+type Epoch {
+  epochId: Int!
+  systemStateVersion: Int
+  protocolConfig: ProtocolConfig
+  referenceGasPrice: U64
+
+  systemParameters: SystemParameters
+  stakeSubsidy: StakeSubsidy
+  validatorSet: ValidatorSet
+  storageFund: StorageFund
+  safeMode: SafeMode
+
+  startTimestampMs: U64
+  endTimestampMs: U64
+
+  # TODO: Identify non-duplicate fields in `EndOfEpochInfo`
+
+  checkpoints(order: Order!): CheckpointConnection
+  transactionBlocks(order: Order!): TransactionBlockConnection
+}
+
+type ProtocolConfig {
+  protocolVersion: Int!
+  # Omitting for brevity
```

Comment:
This is one where it's super annoying to describe manually in the schema if we wanted to describe each field, but straightforward (and easier) to do it that way if we auto-generate the schema from the underlying Rust type.

Reply:
I actually think we don’t want this to be generated out though, because it makes api compatibility harder. E.g. I can only issue a query for a specific piece of protocol config if the server supports it.

Reply:
If we exposed a type like:

```
type ProtocolConfigs {
    protocolVersion: U64
    configs: [ProtocolConfig]
    config(key: String!): String
}
```

Can the `config` query be used multiple times to fetch multiple configs? (In one request).

Reply:
Yeah you can alias the field to do that. We could also have a multi-get interface where you pass an array